src/subword.rs

```diff
@@ -341,6 +352,19 @@ impl<'a> Iterator for NGrams<'a> {
 
         Some(ngram_with_len)
     }
+
+    #[inline]
+    fn size_hint(&self) -> (usize, Option<usize>) {
+        let stop = cmp::min(self.char_offsets.len(), self.max_n);
+        let l = (self.min_n..=stop)
```

Comment:
Seems like there is a closed form?

https://www.wolframalpha.com/input/?i=sum+n-%28j-1%29%2C+j%3Da+to+b

Reply:
I also have a vague recollection that we already used a closed form for this somewhere else.

Reply:
I remember discussing this with Marko while working on the slicing assignment for the Rust course, so mightve been in that context?

Reply:
Could be! In my example solution I overallocated a bit

```
let cap_approx = (n_max - n_min + 1) * seq.len();
```

Maybe I had the exact computation first, but replaced it because it detracted too much from the rest of the solution.

Reply:
I think there are some complications if `min_n > seq.len()` with the closed form from above. Not sure if I misread something but I think this should be it:

https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=7c4ec5c493b63c85ee68787c2e1ead93

This returns 1, although from a sequence of length 2, we wouldn't get any ngrams. Might also be less desirable because there's a bunch of floating point operations rather than simple integer operations. Do you prefer the sum or the approximation? Or should we go for the exact calculation?

Reply:
~~~Rust
        let n = self.char_offsets.len();
        if self.min_n  > n {
            (0, Some(0))
        } else {
            let a = self.min_n as f32;
            let b = cmp::min(n, self.max_n) as f32;
            let l = (0.5*(-1. + a - b)*(a + b - 2. * (1. + n as f32))) as usize;
            (l, Some(l))
        }
~~~

This seems to work but is 10% slower on the subwords bench compared to the sum version. Your approximation is on par with summing.

I'd be in favor with sticking to my initially proposed version since that is probably the easiest to read.

Reply:
If the opportunistic approximation is as fast, I would prefer that. It is constant time and provides better inlining opportunities. These are small allocations anyway, where some over-allocation does not hurt (in contrast to amortized constant appends).